### PR TITLE
internal/servers/controller: Worker failure connection cleanup

### PR DESCRIPTION
**Pull request by vancluever**
This commit adds the support to do the following:

* Mark connections for non-reporting workers as closed. This is the
controller counterpart to the worker functionality (see #1330). This is
written as a scheduled job that does most of the work DB-side, save some
rudimentary checking of individual workers' last update times.

* Works to reconcile states if such a broken controller-worker
connection resumes and a worker reports a connection as connected that
should be disconnected. In this case, the controller will send an update
request, and the worker will honor it and terminate the connection.

* Further refinement of the grace period setting has been added here.
We have converged on the current server "liveness" setting as our
default here, which is half of the previous 30s (15 seconds, in other
words). Additionally, this is now configurable on the controller and
worker side, with the caveat that it's currently impossible to do so in
config as the setting has been untagged in HCL. This is exposed so that
we can run some sophisticated testing scenarios where we skew the grace
period to either the controller or worker to ensure the aforementioned
reconciliation works.

* Some repository functions have been added to support the new
functionality.

### CODE REVIEW

File: internal/cmd/config/config.go
Comment (jefferai) [inline internal/cmd/config/config.go:124]: I don't think we'll ever want this value exposed via config, so rather than put it in here, can you simple set it as a field on the `base.Server` and then it can be toggled in tests from there?

File: internal/servers/controller/controller.go
Comment (jefferai) [inline internal/servers/controller/controller.go:34]: Why did this get changed to 15 from 30? Seems very short.

File: internal/servers/controller/controller.go
Comment (jefferai) [inline internal/servers/controller/controller.go:185]: Rather than copy this code between controllers and workers, please make it a function of `base.Server` that checks the env var and uses that or uses the default (similar to how we set up listeners, controllers, etc.), and sets it in the Server struct. Since both worker and controller have that in their local `Config` object they can share the logic.

File: internal/servers/controller/handlers/workers/worker_service.go
Comment (jefferai) [inline internal/servers/controller/handlers/workers/worker_service.go:89]: This could potentially be thousands or tens of thousands of DB calls. I think we need a different way to handle this. One might be to have an analogue of LookupSessions that allows a WithSessionIds but instead WithConnectionIds, so we just give that list as part of a single query. Another might be to return the current status for each connection along with the session in the list. Both have pros and cons and probably worth a deeper discussion with the DB experts.

Comment (jefferai) [reply internal/servers/controller/handlers/workers/worker_service.go:89]: Or possibly the current state could be returned as part of ListConnectionsByServerId.

File: internal/servers/controller/controller.go
Comment (vancluever) [reply internal/servers/controller/controller.go:34]: This is to put it in line with the server liveness timeout (seen [here](https://github.com/hashicorp/boundary/blob/33929c27e956385fd3fe72b8f672eb67a9701ffc/internal/servers/repository.go#L16)). This is something I wanted to investigate after finishing the work here. Definitely flexible on changing it, but it made sense to me that if a server was not showing up as "alive" on things like `ListServers`, then it'd probably make sense to start disconnecting things from it at around the same time.

File: internal/servers/controller/handlers/workers/worker_service.go
Comment (vancluever) [reply internal/servers/controller/handlers/workers/worker_service.go:89]: :+1: totally fair. Will look at a different way for sure!

**Comment by vancluever** [comment]
@jefferai feedback all makes sense, will work, on applying it tomorrow!

**Comment by vancluever** [comment]
@jefferai this is ready for review again. I'll check in a bit to see if tests passed. 🤞

File: internal/cmd/config/config.go
Comment (jefferai) [reply internal/cmd/config/config.go:124]: I think this can now be removed here and in the Worker block right?

File: internal/servers/controller/testing.go
Comment (jefferai) [inline internal/servers/controller/testing.go:420]: I think this also isn't needed anymore?

File: internal/servers/controller/handlers/workers/worker_service.go
Comment (jefferai) [inline internal/servers/controller/handlers/workers/worker_service.go:198]: If I'm right, for a session that isn't canceled but where individual connections should be canceled, this won't yet work as desired on the worker. The worker ranges through sessions and only takes effect if the (updated) session state is canceling or terminated. I think it needs another case that checks individual connections if the session is still active.

**Comment by jefferai** [comment]
Meta comment -- in the future please don't use branch names with slashes -- our standard is a hyphen instead. This is due to https://github.com/golang/go/issues/38861 (and previous issue https://github.com/golang/go/issues/36902). It's not obvious that this should ever be a problem, and yet... 😀

File: internal/servers/worker/testing.go
Comment (jefferai) [inline internal/servers/worker/testing.go:156]: Same comment, I think this should be using the server var now?

File: internal/session/query.go
Comment (jefferai) [inline internal/session/query.go:300]: I think this comment is describing the opposite of the below check

File: internal/servers/controller/handlers/workers/worker_service.go
Comment (vancluever) [reply internal/servers/controller/handlers/workers/worker_service.go:198]: This is addressed and is actually addressed by the skew cases where we skew in favor of the controller versus the worker. I want to link the work now but will wait until after addressing the rest of the review feedback to ensure that the link is stable. 😛

Comment (vancluever) [reply internal/servers/controller/handlers/workers/worker_service.go:198]: OK, I thought I added this but apparently it's not there. Taking a look now, also need to figure out why the tests are passing if this is indeed the case. The cases where the controller is skewed for should definitely close connections on the worker first, and then cut off the connection when the next status request is made and a change request goes through. Will review this and add the code.

**Comment by vancluever** [comment]
Adding @louisruch to review the job code.

File: internal/servers/controller/controller.go
Comment (louisruch) [inline internal/servers/controller/controller.go:181]: Might be good to merge `main` into `vancluever/session-cleanup` and then `vancluever/session-cleanup` into this branch as main now makes use of this method.

File: internal/servers/controller/session_cleanup_job.go
Comment (louisruch) [inline internal/servers/controller/session_cleanup_job.go:78]: While I don't necessarily disagree with the comment a second does seem aggressive.  Regardless of that please note that each scheduler only requests work once a minute, so you will only truly reach continuously running at scale when multiple controllers are running schedulers.

File: internal/servers/controller/session_cleanup_job.go
Comment (louisruch) [inline internal/servers/controller/session_cleanup_job.go:148]: I am wondering if it might be beneficial to move this call outside of the worker range, and then create a modified query that instead of calling:
```
-- Related to the worker that just reported to us
      server_id = $1
```
does a `server_id in` with a list of worker_ids that we collected in the range which were outside the grace period.

**Review by louisruch — COMMENTED** [review]
Did a first pass, I did not look at tests at all yet.  I will have another look over the weekend once I had some down time - left a couple of comments in the mean time

File: internal/servers/controller/session_cleanup_job.go
Comment (vancluever) [reply internal/servers/controller/session_cleanup_job.go:148]: @louisruch are you referring to having a single query versus more than one? If so, @jefferai and I discussed this and concluded that there would not be enough workers for this to matter.

File: internal/servers/controller/session_cleanup_job.go
Comment (vancluever) [reply internal/servers/controller/session_cleanup_job.go:78]: Yeah, I discovered that after setting this up, hah.

One thing that I'm worried about is that if I set this, and we add a next run time of say, like a minute, that it will possibly be two minutes before the job runs again, as the next scheduled time may not come up before the next job run happens.

Is there any other/better way to guarantee that this job runs once per minute? The actual grace period is 15 seconds, so we want to make absolute sure that it runs as soon as possible to ensure connections are closed as soon as they can be.

File: internal/servers/controller/session_cleanup_job.go
Comment (louisruch) [reply internal/servers/controller/session_cleanup_job.go:148]: That is what I was suggesting.  Are we confident with the claim there won’t be enough workers to worry about?  This is something we can revisit in the future though, just felt like a relatively easy optimization.

Comment (vancluever) [reply internal/servers/controller/session_cleanup_job.go:148]: @jefferai thoughts? Do we have any data on what our upper limit on scale right now is for workers?

@louisruch I think that in terms of orders of magnitude, running a query for every _worker_ is probably much safer than running a query for, say, every _connection_. Probably within the acceptable range, but it would be good to know what our expect scale is for workers before saying 100% for sure.

File: internal/servers/controller/session_cleanup_job.go
Comment (vancluever) [reply internal/servers/controller/session_cleanup_job.go:78]: Discussed this with @louisruch offline and agreed on the fact that the 1 second interval should be good for now to ensure that the job always runs within a particular controller's minimum scheduler interval (1 minute). This will be increasingly variable as the number of controllers go up (and assuming that they were _not_ all started at around the same time), but it helps cover cases where there will be a small amount of controllers where the possibility of the actual next run time being a full minute is higher.

File: internal/servers/controller/session_cleanup_job.go
Comment (jefferai) [reply internal/servers/controller/session_cleanup_job.go:148]: Given that the database holds the last updated time, I wonder if we even need to calculate each worker's grace individually. Think about a supermarket where they have a sign like, if the birth date is past today in 1974, that person can buy booze. I wonder if we really can and ought to use the DB for this. Basically, just tell the DB, if the last status time is before now minus 15 seconds, close all connections. I think this could generate the list and close things in a single CTE in a single transaction.

File: internal/cmd/base/servers.go
Comment (jimlambrt) [inline internal/cmd/base/servers.go:645]: I'm curious about the need/choice to set this via an envvar.  It's not a typical pattern in Boundary.  Is there some specific use-case that requires envvar support?

File: internal/servers/controller/session_cleanup_job.go
Comment (jimlambrt) [reply internal/servers/controller/session_cleanup_job.go:148]: @jefferai you're suggestion would seem to scale better and have the benefit of eliminating some race conditions as well.

Comment (vancluever) [reply internal/servers/controller/session_cleanup_job.go:148]: Alright, will move this to a single query. Incidentally, this was the original thought (when I was looking at table changes to make this happen), but it was scrapped when we decided that multiple queries was fine.

I'm sure we can do it without additional table changes though, so don't take that as me saying that we need to do that again.

File: internal/cmd/base/servers.go
Comment (vancluever) [reply internal/cmd/base/servers.go:645]: @jefferai ^^^

File: internal/servers/controller/controller.go
Comment (vancluever) [reply internal/servers/controller/controller.go:181]: Going to rebase this to main post-release actually, as that "long-running" branch is not being used anymore.

File: internal/cmd/base/servers.go
Comment (jefferai) [reply internal/cmd/base/servers.go:645]: @jimlambrt The issue here is that really we want this to be a database value so that it's not a per-controller configuration value that might be different on different controllers. So I'm hesitant to add it to config as a stop-gap because then we'd have to remove it later and force users to switch over.

What this really needs is the previously discussed attributes/defaults system -- the same thing that default worker behavior on targets is waiting on, as well as synchronized user preferences. The idea for the moment is it's an env var that is unpublished/undocumented that acts as a relief valve for anyone having issues with the implementation, but that officially we only have a static value right now.

Comment (jimlambrt) [reply internal/cmd/base/servers.go:645]: @jefferai and I chatted about this... @vancluever can we add a comment that makes it explicit that this is a temporary stopgap measure?  This will make it clear to everyone that the envvar is not a permanent solution and we  will definitely deprecated it in the future in favor of a holistic approach to attributes/defaults system.   ty!

File: internal/servers/controller/controller.go
Comment (vancluever) [reply internal/servers/controller/controller.go:191]: This was done last week (as discussed with @louisruch offline so that we could resolve some job throttling issues).

File: internal/cmd/base/servers.go
Comment (vancluever) [reply internal/cmd/base/servers.go:660]: Done!

File: internal/servers/controller/session_cleanup_job.go
Comment (vancluever) [reply internal/servers/controller/session_cleanup_job.go]: This is now done!

File: internal/servers/controller/handlers/workers/worker_service.go
Comment (vancluever) [reply internal/servers/controller/handlers/workers/worker_service.go:198]: Resolving old fixed conversation!

File: internal/servers/worker/status.go
Comment (jefferai) [inline internal/servers/worker/status.go:75]: Another way to do this which feels _to me_ to be more canonical but very well may not be is to have a `select` statement on `ctx.Done()` and `time.After(time.Second)`. This will work, but may be worth looking into.

File: internal/servers/controller/session_cleanup_job.go
Comment (louisruch) [inline internal/servers/controller/session_cleanup_job.go:80]: Its fine if you want to keep it like this, but with the Vault Credential Cleanup work I set both `Completed` and `Total` to the number of affected rows in the single query https://github.com/hashicorp/boundary/blob/main/internal/credential/vault/jobs.go#L930.  
This gives us a possibly helpful future record (outside of logs) of how many creds we to cleanup in a job.  We could similarly set a field to `len(results)` in `Run`

File: internal/servers/repository.go
Comment (louisruch) [inline internal/servers/repository.go:82]: I am not clear how `liveness` will ever `<= 0`.  Doesn't the `if` statement above prevent this?  Unless of course we change the `DefaultLiveness` to `0` at some point but that should never happen anyway...

File: internal/servers/repository.go
Comment (louisruch) [inline internal/servers/repository.go:83]: we have a new domain helper function Mike added for Vault that you could use here as:
`update_time > wt_add_seconds_to_now(?)` and then add `-1 * liveness.Seconds()` to the args.

File: internal/servers/worker/status.go
Comment (louisruch) [inline internal/servers/worker/status.go:75]: Seems to be a good use-case for using a [Cond](https://golang.org/pkg/sync/#Cond), where the worker can just broadcast every time it performs an update to the `lastStatusSuccess` instead of polling every second.  Ok keeping it like this as well if that causes a larger refactor that seems.

File: internal/servers/worker/testing.go
Comment (louisruch) [inline internal/servers/worker/testing.go:214]: Is this needed?  Doesn't `SetStatusGracePeriodDuration` below handle both the zero and non-zero case?

File: internal/session/repository_connection.go
Comment (louisruch) [inline internal/session/repository_connection.go:185]: We should add `defer rows.Close()` here

File: internal/session/repository_connection.go
Comment (louisruch) [inline internal/session/repository_connection.go:260]: We should add defer rows.Close() here

File: internal/session/query.go
Comment (louisruch) [inline internal/session/query.go:300]: Could also use the `wt_add_seconds_to_now` function here, I feel it reads better but as you like.  Could also create a new `wt_subtract_seconds_from_now` if you want to avoid sending the arg with a `-1*`

**Review by louisruch — COMMENTED** [review]
Couple of comments but this is really looking great!

File: internal/servers/worker/status.go
Comment (vancluever) [reply internal/servers/worker/status.go:75]: Let's leave this like it is right now.  I think adjusting the status check inter-communication in such a way would be a bit more impacting than just adjusting things here, in addition to me not necessarily understanding sync.Cond well (and I haven't seen many good examples of its use). Could be something we entertain in the future.

File: internal/servers/worker/testing.go
Comment (vancluever) [reply internal/servers/worker/testing.go:214]: Yeah, it's one or the other. Probably a relic of maintenance, will fix either way.

**Comment by vancluever** [comment]
@louisruch comments addressed!

**Comment by vancluever** [comment]
Also just rebased off of main to fix conflicts.

File: internal/servers/repository.go
Comment (vancluever) [reply internal/servers/repository.go:82]: Reverted this as we actually use a negative number here to disable the default liveness setting.

Comment (vancluever) [reply internal/servers/repository.go:82]: See https://github.com/hashicorp/boundary/blob/ebc0b96b0b7dc989202f8fba255de79e251f2164/internal/servers/options.go#L40-L47

Comment (louisruch) [reply internal/servers/repository.go:82]: ty

File: internal/servers/controller/testing.go
Comment (louisruch) [inline internal/servers/controller/testing.go:652]: Not sure if you wanted to implement the same suggestion Jeff gave re turning this into a for select

**Review by louisruch — DISMISSED** [review]
Not sure if @jefferai wanted to review this again before merging but this LGTM

File: internal/servers/controller/testing.go
Comment (vancluever) [reply internal/servers/controller/testing.go]: Should be done now!

**Review by louisruch — APPROVED** [review]
LGTM

File: internal/servers/controller/testing.go
Comment (jefferai) [inline internal/servers/controller/testing.go:645]: Why are these two functions separate? One just calls the other -- why not just have a single function?

I looked and don't see it used anywhere else -- even tests only use the exported version.

File: internal/servers/controller/testing.go
Comment (jefferai) [inline internal/servers/controller/testing.go:675]: Don't you want to return `true` in all of these cases? Finding a bad value is obviously a bad thing, but it doesn't mean that the value for the worker ID passed in is bad. You could have perfectly legitimate data for the worker ID we care about but be ignoring it by cutting out early on an error for some other value.

File: internal/servers/controller/testing.go
Comment (jefferai) [inline internal/servers/controller/testing.go:698]: Similar question -- if the data for the worker we're interested in is good, why return an error here? It's good to surface that something is wrong, but not ideal to not perform the intended function if that data is fine.

File: internal/servers/worker/status.go
Comment (jefferai) [inline internal/servers/worker/status.go:209]: This could be simplified to use `connId` from the line above.

File: internal/servers/worker/status.go
Comment (jefferai) [inline internal/servers/worker/status.go:271]: Given it's only two places and only a single line for the function, my personal preference would be to just duplicate the logging lines here and below. To me it's a bit less mental energy to just read the line here than have to remember or jump down to the function to see what it's doing.

Comment (jefferai) [reply internal/servers/worker/status.go:271]: Take it or leave it, just my $0.02.

File: internal/session/query.go
Comment (jefferai) [inline internal/session/query.go:311]: Would it be better to do something like `state not in ('closed')`? That way if we add more states prior to `closed` then it still works.

File: internal/session/repository_connection.go
Comment (jefferai) [inline internal/session/repository_connection.go:181]: It'd be nicer to use ScanRows below -- it's not in the writer interface right now but it could be added to the interface with no further change to code as the underlying db object already has that function.

File: internal/servers/controller/testing.go
Comment (vancluever) [reply internal/servers/controller/testing.go:645]: This was to avoid some issues with short-circuiting on context timeouts in the `select` at the top of the inner function's `for` loop. I've fixed this and by just wrapping the select in a function so we can `break` properly and the fix will be in the next PR.

File: internal/servers/controller/testing.go
Comment (vancluever) [reply internal/servers/controller/testing.go:675]: This is for type asserting that the key is a string and the value is a `time.Time`, just basically validating the data integrity in the `sync.Map`. It's more about ensuring that if there are unexpected values in `WorkerStatusUpdateTimes`, it probably means that there is a problem somewhere else and we should probably investigate it more.

I think the other thing we could do is just type assert without validation, given that this function is only designed for testing currently. Will leave for now, but let me know your thoughts.

File: internal/servers/controller/testing.go
Comment (vancluever) [reply internal/servers/controller/testing.go:698]: See the [above reply](https://github.com/hashicorp/boundary/pull/1340#discussion_r670884050). Again, fine with just removing it completely and letting it panic instead, but I don't think we just continue given that we expect the data in the map to be a very specific structure. That is, unless I'm missing something and there would be data in `WorkerStatusUpdateTimes` where the key/value would not be `string`/`time.Time`.

Comment (jefferai) [reply internal/servers/controller/testing.go:698]: I don't think we need to panic; you'd skip the bad data and move on to the next piece of data in the range.

That said this is in testing code so especially in that case assuming it will be good and erroring on very unexpected data is probably fine.

**Pull request merged**
